### ISSUE TITLE
Please Clarify How Bindings Work, Trying to Port Pieces to Swift?

### ISSUE THREAD

**Issue by KTRosenberg** [phoboslab/Ejecta#674]
Hello. I realize the project is rather old at this point, but it's probably the best example of using JavaScriptCore for anything. Would it be alright if I ask for clarification and assistance with figuring out the correct bindings?

I'm currently trying to use JavaScriptCore in Swift in an attempt to enable remote debug scripting, hot reloading, and hopefully scriptable Metal (much like how WebGL was implemented here.) However, I'm not sure I can really reuse Ejecta, so that's why I'm only trying to create a simplified subset of what I need in Swift.

However, I am having a lot of trouble figuring how to manage communication between Swift and JavaScript.

For some context, what I'd like to do is to allocate and share memory in the form of typed arrays so that JavaScript can write commands into the buffer and Swift can read/respond to those commands. I assume it makes sense for Swift to allocate this array if I want to reload the JavaScript code at any point. In addition, I will probably do something similar to getImage() and the WebSocket communication.

The actual problem is that I can't seem to figure how to return data from Swift to JavaScript. Whenever I set a bound a Swift function's return value to a `JSValueRef`, JavaScript doesn't see that function (`undefined`). Swift has to resort to calling conventions similar to Objective C, so I don't think what I want to do is impossible.

I'll post my code followed by links to the pieces of Ejecta that I think are relevant:

### Swift
```swift
@objc protocol JSMemoryExports: JSExport {
    static func alloc(_ byte_count: Int) -> JSValueRef
}

class JSMemory: NSObject, JSMemoryExports {
    class func alloc(_ byte_count: Int) -> JSValueRef {
        let jsContext = JS_Controller.js.ctx!
 
//        var jsref : UnsafeMutablePointer<JSValueRef?>?

        return JSObjectMakeTypedArray(
            jsContext.jsGlobalContextRef!, 
            kJSTypedArrayTypeUint8Array, 
            byte_count, 
            nil
        )
}

...

jsContext["mem"] = JSMemory.self
...
jsContext.evaluateScript(jsSourceContents)

```
### JavaScript
```js
const buf = new Uint8Array(mem.alloc(8))
```
### Result
`JS Exception: TypeError: mem.alloc is not a function. (In 'mem.alloc(8)', 'mem.alloc' is undefined)`

If I have no return value, it's fine, but I'm not sure how to achieve what you did with the WebGL API. Admittedly, the macros and Objective-C syntax are rather hard to follow.

### Relevant Ejecta Code
It looks like the function [here](https://github.com/phoboslab/Ejecta/blob/399a10f446a878baa95d5cffad6ec2d602acf71a/Source/Ejecta/EJCanvas/WebGL/EJBindingCanvasContextWebGL.m#L717)
and specifically this part:
```c
ret = JSObjectMakeTypedArray(ctx, kJSTypedArrayTypeInt32Array, sizeof(GLint) * 4, NULL);
```
Is creating a typed array allocated in Objective-C and returned to JavaScript, which is exactly what I'm trying to do. The return type is also JSValueRef I think, so I'm following this correctly too.

Yet looking into the `EJ_BIND_FUNCTION` macros, I am unsure of what you're actually doing to bind the functions. I cannot follow.

Of course this project is several years old, but would you be willing to help me figure what the proper way of binding these functions is? One thing I also didn't consider was how I'd free this memory or avoid retention cycles, but first things first, I need something working.

I WAS considering using Ejecta by the way, but if I am unsure of how to modify it, then that might not work so well.
Thank you for your time.

**Comment by phoboslab** [comment]
I'm sorry, I don't think I can help you much. I don't know anything about Swift and very little about JSC's newer APIs.

Are you sure the APIs supports exporting static class methods? A quick google search at least returned an article that explains how to bind a Swift class and instantiate it in JS - with instance methods and all: https://nshipster.com/javascriptcore/

Just for reference, the `EJ_BIND_*` macros define a static objc class method that returns a function pointer to the actual implementation (the body of the macro), that is later collected by [EJClassLoader](https://github.com/phoboslab/Ejecta/blob/master/Source/Ejecta/EJClassLoader.m#L126-L156) and handed over to JSCs `JSClassCreate()`. EJClassLoader (and all the rest of Ejecta) uses JSCs older C API, listed [here under Legacy JavaScriptCore APIs](https://developer.apple.com/documentation/javascriptcore?language=objc), not the newer ObjC API.

**Comment by KTRosenberg** [comment]
Hmmm, yes I'm sure that static functions are supported. The link you have there shows the following:

```
// Protocol must be declared with `@objc`
@objc protocol PersonJSExports: JSExport {
    var firstName: String { get set }
    var lastName: String { get set }
    var birthYear: NSNumber? { get set }

    var fullName: String { get }

    // Imported as `Person.createWithFirstNameLastName(_:_:)`
    static func createWith(firstName: String, lastName: String) -> Person
}
```

That is basically what I am doing. I've found several examples online of passing back and forth custom classes, but none showing how to pass raw data such as a TypedArray or JSValueRef, which is what seems to cause issues.

So now I'm wondering what the proper way of passing byte data back and forth without copying is. It's admittedly frustrating not to see any real examples.

**Comment by KTRosenberg** [comment]
Just for an update, I've figured out how to mix the old C API with the new more limited Swift-specific APIs. I haven't yet made sure I'm not leaking memory, but it looks like I've found what I needed, hopefully.

In case you ever wanted to know:

```swift
@objc protocol JSMemoryExports: JSExport {
    // note that I'm returning an optional
    static func Uint8ArrayMake(_ count : JSValue) -> JSValue?
}

class JSMemory: NSObject, JSMemoryExports {
    //https://www.raywenderlich.com/7181017-unsafe-swift-using-pointers-and-interacting-with-c    
    class func UInt8ArrayMake(_ count : JSValue) -> JSValue? {
        guard !count.isUndefined && !count.isNull else {
            return nil
        }
        
        let ref : JSValueRef = JSObjectMakeTypedArray(
            JS_Controller.js.ctx.jsGlobalContextRef!,
            kJSTypedArrayTypeUint8Array,
            Int(count.toInt32()),
            nil
        )!
          
        let ptr = JSObjectGetTypedArrayBytesPtr(JS_Controller.js.ctx.jsGlobalContextRef!, ref, nil)
        return JSValue(jsValueRef: ref, in: JS_Controller.js.ctx)    
    }
}
```
